### DISCUSSION TITLE
Domain name tracking

### DISCUSSION THREAD

**Discussion by subdriven** [sys4/netbox-plugin-dns#53] [Ideas]
I thought a discussion thread might be a better place for this than an issue. I was thinking this plugin would be a good place to track not just domain name server information, but actual domain names as well. My thought is a similar page to _records_ (maybe _names_?) where you can track the domain name, registrar, renewal date, and maybe internal owner among other things. Obviously the _comments_ section could be used to document items more nuanced to each organization but as an organization, we have so many domain names for various things, various departments, etc. that keeping it in the dreaded spreadsheet is a pain. 

There could be possibility to link the _records_ to the _names_ in various ways I'm sure - I haven't really thought it all out yet but wanted to float it before going too far down a rabbit hole.

**Comment by kollross** [comment]
I personally would be all for this.  I'm in a similar situation, where we purchase domains for specific projects, then come renewal time, have to spend time tracking down who the owner of that domain/project was to find out if they want to renew it.   Right now it's tracked in a wiki, but Netbox seems like a more logical location.  

I say this fully admitting, that I lack the skill set to implement any such features.

**Reply by subdriven** [reply]
I _might_ be able to slap something together eventually, but programming is not my strength. Seems like it could be fairly straight forward if the specifications were exact enough.

**Comment by peteeckel** [comment]
Actually all of this is possible right now with NetBox standard functionality.

* A 'domain name' is directly equivalent to a zone, so there's no need for any additional model or field.
* 'registrar', 'renewal date' and 'internal owner' can easily be implemented as custom fields on `Zone` objects.

That said, it might make sense to add some organisational fields to Zone objects. A `Registrar` model would make it possible to select registrars from objects in the database, making it easier to enter new objects and maintain integrity. The 'internal owner' could be mapped to a `Tenant` (that feature suggestion was already made a long time ago, but never made it to the top of the list due to lack of interest).

For the time being, custom fields should do the trick, though.

**Reply by subdriven** [reply]
Well when you put it that way :-)

That makes sense to me. I can try filling this out with a few domains we have but I do think adding some models to the Zone objects would make some sense. I see where mapping the internal owner would make sense, however in our use case, we don't have our departments listed as tenants besides the tech teams.

**Reply by peteeckel** [reply]
Just try with custom fields, and when you get a feeling for what you really need in terms of functionality not possible with that approach that would be a good start for planning enhancements.

I was never really convinced tenancy for zones and nameservers (and much less records or views) makes much sense, but if a good point exists for adding the feature I wouldn't mind. It's not rocket science, but I try to keep things as free from unnecessary complexity as possible.

**Reply by subdriven** [reply]
Is there a plan to link this plugin with IPAM? I could see where internal systems with reverse lookups to IP addresses and hosts would make some sense.

**Reply by kollross** [reply]
There is a script in the scripts directory  that updates the dns_hostname field in IPAM from the plugin.

**Reply by peteeckel** [reply]
Hi @subdriven, 

> Is there a plan to link this plugin with IPAM? I could see where internal systems with reverse lookups to IP addresses and hosts would make some sense.

Bottom line: Yes, it would make sense. I wouldn't say there is a plan to implement this, but there have been some discussions online and offline how to manage it, and so far none lead to a feasible plan. 

The problem is that although it may seem trivial to map IPAddress objects and their `dns_name` properties to address and pointer records in DNS, it by no means is. Addresses can be non-unique (with or without the presence of VRFs), since NetBox is very permissive in that respect, and DNS zones and records can be the same if you consider Views. To make things worse, VRFs do not generally map to Views. There are some more issues to be considered, and all in all it's surprisingly difficult to come up with a reasonably general mechanism.

For specific cases it's always possible to create a solution based on the scripts @kollross mentioned, though. From personal experience I would rather fill in `dns_name` for IPAddress objects from DNS (be it only for the better validation of NetBox DNS host and domain names), but YMMV.

**Reply by subdriven** [reply]
@peteeckel yes, I can see that it gets really complicated really quickly. Perhaps I will stick with what we have and put some custom fields where necessary. Thank you for your help and replies - really appreciate it!

**Reply by peteeckel** [reply]
We have a simple and working solution now - not fully automated, but it's there, it's usable and it's probably useful for your needs.

Essentially, the new 'IPAM Coupling' feature contributed by @jean1 can be used to link IP addresses to NetBox DNS address records by specifying a recore name - usually the host name - and a zone for each address. NetBox DNS will then create an address record for it and link the new address record to the IP address in a way that changes to one of them (changing the address or renaming the host or zone) are reflected in the other one (e.g. by updating 'DNS Name' for the IP address, the value of the address record or deleting the address record when an IP address is removed).

Some additions to the UI have been added as well so that you can see the IP address in the record's detail view and vice versa.

At the very least it is a good start and might take away the pain for some users. Please give it a try and tell us what you think!

**Comment by subdriven** [comment]
Yeah, I think your idea should work for our needs. No reason to expand that I can see. 
![image](https://github.com/peteeckel/netbox-plugin-dns/assets/5957345/518259d9-03a6-44ed-9e35-d359eb4b24fa)

**Comment by peteeckel** [comment]
Just for the record - I just added the tenancy feature for NetBox DNS in version 0.19.0. It wasn't exactly complicated and it could be a step towards some other stuff that recently popped up in a discussion - not sure if it will be the right way to go, but it was cheap and works, so here we go.

Tenancy can be added to all models. The only notable exception are managed records that can't actually be edited manually. I could make them inherit tenancy from their associated address records (in the case of PTR records) or the zone they belong to (in the case of SOA or NS records), but I'm not sure if that makes sense.

**Comment by peteeckel** [comment]
Short update: I just added the feature to assign registrars and various contacts (registrant, tech-c, admin-c, billing-c) to NetBox DNS 0.21.0.

If the custom field solution does not fully cover your use cases you might be interested in trying the latest version.

As always, feedback ist welcome!
